src/odm_sharing/private/cons.py

```diff
@@ -23,37 +34,186 @@ def _write_table_to_db(db: sa.engine.Engine, table: str, df: pd.DataFrame
+    '''Finds bool columns in an excel sheet'''
+    result = defaultdict(set)
+    formula_wb = xl.load_workbook(path, read_only=True, data_only=False)
+    for data_sheet, formula_sheet in zip(wb, formula_wb):
+        table = data_sheet.title
+        if table not in included_tables:
+            continue
+        columns = seq(next(data_sheet.iter_rows(max_row=1)))\
+            .map(lambda header: header.value)\
+            .list()
+        bool_col_ixs = _get_excel_bool_col_ixs(data_sheet, formula_sheet)
+        bool_cols = seq(bool_col_ixs).map(lambda i: columns[i]).set()
+        result[table] = bool_cols
+    return result
 
 
 def _connect_excel(path: str, table_whitelist: Set[str]) -> Connection:
-    '''copies file data to in-memory db
+    '''copies excel file data to in-memory db
+
+    :returns: a connection to the db
+
+    :raises OSError:
+    '''
+    # XXX: we must NOT change the data (#31).
+
+    # XXX: We can NOT use Pandas to import Excel files, since it may convert
+    # booleans to float when the first column cell value isn't a valid boolean.
+    # This happens even with `dtype=str`. Invalid booleans happen because
+    # we must allow empty cells and NA values, and we can't normalize the data
+    # (#31). Pandas uses openpyxl under the hood as its Excel backend, and we
+    # can use it directly to avoid this issue.
+
+    # XXX: We will NOT use `dtype=str` when converting the imported data to
+    # Pandas-dataframes, since there's no need at this point.
+
+    # XXX: After import, boolean values will be 0/1, which we'll have to
+    # convert back (using previously detected bool columns) to FALSE/TRUE
+    # before returning the data to the user. This happens in the sharing
+    # module.
 
-    :raises OSError:'''
     logging.info('importing excel workbook')
-    db = _create_memory_db()
-    xl = pd.ExcelFile(path)
-    included_tables = set(map(str, xl.sheet_names))  # XXX: may be int
+
+    # load excel file
+    wb = xl.load_workbook(path, read_only=True, data_only=True)
+    sheet_names = seq(wb).map(lambda sheet: sheet.title).list()
+    included_tables = set(sheet_names)
     if table_whitelist:
         included_tables &= table_whitelist
-    for table in included_tables:
-        df = xl.parse(sheet_name=table)
-        _write_table_to_db(db, table, df)
-    return db
+    bool_cols = _get_excel_bool_cols(path, wb, included_tables)
+
+    # convert to dataframes
+    dfs = {}
+    for sheet in wb:
+        table = sheet.title
+        if table in included_tables:
+            row_iter = sheet.values
+            columns = list(next(row_iter))
+            df = pd.DataFrame(row_iter, columns=columns)
+            dfs[table] = df
+
+    # write to db
+    db = _datasets_to_db(dfs)
+    return Connection(db, bool_cols)
 
 
 def _connect_db(url: str) -> Connection:
     ''':raises sa.exc.OperationalError:'''
-    return sa.create_engine(url)
+    return Connection(sa.create_engine(url), defaultdict(set))
```

Comment:
Shouldn't this get all bool columns from the URL?

Reply:
when connecting to a db we don't have to import anymore, since we're running the queries directly on the source data.

Reply:
Ok, so if the datasource is a database, then the bool columns parameter does not have to be filled in?

Reply:
Is this planned for this version? If not we should add a warning in the documentation about it.

Reply:
I'll look into how feasible it is and get back to you